Refactoring of the RMT component in the kernel

sandervrijders opened the pull request:
This refactors the RMT. Fixes several bad locking and possible memory leaks. All files related to RMT now also pass the check with checkpatch.pl. Henceforth any RMT changes should be checked with it as well. 

Maintainer: @miqueltarzan for everything RMT related, @lbergesio for dt-utils.c

lbergesio commented on linux/net/rina/rmt.c:
```diff
@@ -53,518 +53,545 @@
+	rcu_read_unlock();
+
+	n1_port_destroy(n1_port);
+	spin_unlock_irqrestore(&n1_port->lock, flags);
```
function ```n1_port_cleanup``` does 
```
n1_port_destroy(n1_port);
spin_unlock_irqrestore(&n1_port->lock, flags);
```
while ```n1_port_destroy``` already frees ```n1_port```

sandervrijders replied:
Good catch. Fixing that in a new commit. Didn't get errors about it (luckily?).

lbergesio commented on linux/net/rina/dt-utils.c:
```diff
@@ -953,7 +953,7 @@ int common_efcp_pdu_send(struct efcp * efcp,
 
 	/* Remote flow case */
 	if (pci_source(pci) != pci_destination(pci)) {
-	        if (rmt_send(rmt, pci, pdu)) {
```
The signature of ```rmt_send``` included the PCI just to avoid having to retrieve it again, but ok, I was about to change this once.

sandervrijders replied:
Yeah, there's no overhead in retrieving it once more. I prefer to keep the function signatures minimal.

lbergesio commented on linux/net/rina/rmt.c:
```diff
@@ -667,466 +692,459 @@ int rmt_sdup_config_set(struct rmt *         instance,
-int rmt_config_set(struct rmt *        instance,
-                   struct rmt_config * rmt_config)
+int rmt_config_set(struct rmt *instance,
+		   struct rmt_config *rmt_config)
```
missing space in line 702 here

lbergesio replied:
missing space in line 702 here

lbergesio commented on linux/net/rina/rmt.c:
```diff
@@ -667,466 +692,459 @@ int rmt_sdup_config_set(struct rmt *         instance,
-                                 struct pdu *         pdu)
+static int n1_port_write(struct rmt *rmt,
+			 struct rmt_n1_port *n1_port,
+			 struct pdu *pdu)
```
missing spaces in lines 739 and 740

lbergesio commented:
What are the important changes in addition to all the formatting? There was a point where I could not follow because all the blocks were in red/green

sandervrijders commented:
You can always pull the file directly and go through it. I redid the locking from the bottom up. Which resulted in a cleaner workflow, more easily understandable, and got rid of bad locking and several memory leaks.

dstaesse commented:
ACK. Tested with non-blocking I/O #713.
```
4116(1441282160)#traffic-generator (INFO): Port    2:      7668 SDUs (    11271960 bytes) in       100001 us => 76679.2266 p/s,  901.7478 Mb/s
4116(1441282160)#traffic-generator (INFO): Port    2:      7643 SDUs (    11235210 bytes) in       100001 us => 76429.2344 p/s,  898.8078 Mb/s
4116(1441282160)#traffic-generator (INFO): Port    2:      7649 SDUs (    11244030 bytes) in       100001 us => 76489.2344 p/s,  899.5134 Mb/s
4116(1441282160)#traffic-generator (INFO): Port    2:      7659 SDUs (    11258730 bytes) in       100001 us => 76589.2344 p/s,  900.6894 Mb/s
4116(1441282160)#traffic-generator (INFO): Port    2:      7653 SDUs (    11249910 bytes) in       100001 us => 76529.2344 p/s,  899.9838 Mb/s
4116(1441282160)#traffic-generator (INFO): Port    2:      7654 SDUs (    11251380 bytes) in       100001 us => 76539.2344 p/s,  900.1014 Mb/s
4116(1441282160)#traffic-generator (INFO): Port    2:      7651 SDUs (    11246970 bytes) in       100001 us => 76509.2344 p/s,  899.7486 Mb/s
4116(1441282161)#traffic-generator (INFO): Port    2:      7648 SDUs (    11242560 bytes) in       100002 us => 76478.4766 p/s,  899.3868 Mb/s
```
over 920Mb/s internal queues start clogging:
```
4116(1441281970)#traffic-generator (INFO): Port    2:      7484 SDUs (    10851800 bytes) in       100037 us => 74812.3203 p/s,  867.8229 Mb/s
4116(1441281971)#traffic-generator (INFO): Port    2:      7516 SDUs (    10898200 bytes) in       100002 us => 75158.5000 p/s,  871.8386 Mb/s
4116(1441281971)#traffic-generator (INFO): Port    2:      7463 SDUs (    10821350 bytes) in       100101 us => 74554.6953 p/s,  864.8345 Mb/s
4116(1441281971)#traffic-generator (INFO): Port    2:      2974 SDUs (     4312300 bytes) in       100001 us => 29739.7012 p/s,  344.9806 Mb/s
4116(1441281971)#traffic-generator (INFO): Port    2:      1122 SDUs (     1626900 bytes) in       100001 us => 11219.8877 p/s,  130.1507 Mb/s
4116(1441281971)#traffic-generator (INFO): Port    2:      1099 SDUs (     1593550 bytes) in       100001 us => 10989.8906 p/s,  127.4827 Mb/s
4116(1441281971)#traffic-generator (INFO): Port    2:      1107 SDUs (     1605150 bytes) in       100001 us => 11069.8896 p/s,  128.4107 Mb/s
4116(1441281971)#traffic-generator (INFO): Port    2:      1028 SDUs (     1490600 bytes) in       100001 us => 10279.8975 p/s,  119.2468 Mb/s
4116(1441281971)#traffic-generator (INFO): Port    2:      1021 SDUs (     1480450 bytes) in       100001 us => 10209.8975 p/s,  118.4348 Mb/s
4116(1441281971)#traffic-generator (INFO): Port    2:      1060 SDUs (     1537000 bytes) in       100001 us => 10599.8945 p/s,  122.9588 Mb/s
```

however, good improvement!

dstaesse commented:
@miqueltarzan, @lbergesio, can you ACK this?

edugrasa commented:
@miqueltarzan is still testing it, wants to make sure that some use cases that complicated the locking  of the RMT are still supported without kernel bangs (like destroying N-1 flows while there is data going through them)

edugrasa commented:
Got a couple of errors (error 1 is not on pristine-1.3, error 2 I've not been able to reproduce it, but not 100% sure). 

Error 1 (with usual normal over eth configuration):  Due to me forgetting to load the shim-eth-vlan module, creation of shim-eth IPCP failed, when the IPCM proceeded to create the normal IPCP the kernel banged:

  229.816070] BUG: unable to handle kernel NULL pointer dereference at           (null)
[  229.820052] IP: [<ffffffff8170ecb0>] __mutex_lock_slowpath+0xb0/0x1f0
[  229.820052] PGD 1b5d0067 PUD 1aa14067 PMD 0 
[  229.820052] Oops: 0002 [#1] SMP 
[  229.820052] Modules linked in: nfsd auth_rpcgss oid_registry nfs_acl nfs lockd fscache sunrpc normal_ipcp rina_personality_default 8021q garp stp llc psmouse i2c_piix4 pcspkr serio_raw hid_generic floppy usbmouse usbhid hid
[  229.820052] CPU: 0 PID: 3165 Comm: ipcm Not tainted 3.15.0 #7
[  229.820052] Hardware name: Xen HVM domU, BIOS 4.1.4 08/17/2014
[  229.820052] task: ffff88001a916100 ti: ffff88001b5cc000 task.ti: ffff88001b5cc000
[  229.820052] RIP: 0010:[<ffffffff8170ecb0>]  [<ffffffff8170ecb0>] __mutex_lock_slowpath+0xb0/0x1f0
[  229.820052] RSP: 0018:ffff88001b5cfdc0  EFLAGS: 00210286
[  229.820052] RAX: 0000000000000000 RBX: ffff88002fd8e850 RCX: 0000000000000000
[  229.820052] RDX: 0000000000000000 RSI: ffff88001a916100 RDI: ffff88002fd8e854
[  229.820052] RBP: ffff88001b5cfe10 R08: ffff88001b5cc000 R09: 0000000000000001
[  229.820052] R10: 0000000000000000 R11: 0000000000000001 R12: ffff88002fd8e854
[  229.820052] R13: ffff88001a916100 R14: 00000000ffffffff R15: ffff88002fd8e858
[  229.820052] FS:  0000000000000000(0000) GS:ffff88003fc00000(0063) knlGS:00000000f6da4940
[  229.820052] CS:  0010 DS: 002b ES: 002b CR0: 000000008005003b
[  229.820052] CR2: 0000000000000000 CR3: 000000001a961000 CR4: 00000000001006f0
[  229.820052] Stack:
[  229.820052]  ffff88002fd8e858 0000000000000000 ffff88002caa2580 ffff88003d01c0c0
[  229.820052]  ffff88003d01c0c0 ffff88002fd8e850 ffff88002fd8e850 ffff88002caa2580
[  229.820052]  ffff88003d01c0c0 ffff88003d01c0c0 ffff88001b5cfe28 ffffffff8170ee0f
[  229.820052] Call Trace:
[  229.820052]  [<ffffffff8170ee0f>] mutex_lock+0x1f/0x30
[  229.820052]  [<ffffffffa00a9e8b>] rina_component_fini+0x1b/0x50 [rina_personality_default]
[  229.820052]  [<ffffffffa00a203c>] rmt_destroy+0x6c/0xb0 [rina_personality_default]
[  229.820052]  [<ffffffffa00a2d35>] rmt_create+0x135/0x170 [rina_personality_default]
[  229.820052]  [<ffffffffa00d25d8>] normal_create+0x128/0x340 [normal_ipcp]
[  229.820052]  [<ffffffffa00a905b>] kipcm_ipcp_create+0xbb/0x1a0 [rina_personality_default]
[  229.820052]  [<ffffffffa00a98bd>] default_ipc_create+0x3d/0x50 [rina_personality_default]
[  229.820052]  [<ffffffff816f893e>] SyS_ipc_create+0xbe/0x140
[  229.820052]  [<ffffffff8171b9ac>] sysenter_dispatch+0x7/0x21
[  229.820052] Code: 00 8b 03 85 c0 78 09 31 c0 87 03 83 f8 01 74 74 48 8b 43 10 4c 8d 7b 08 48 89 63 10 41 be ff ff ff ff 4c 89 3c 24 48 89 44 24 08 <48> 89 20 4c 89 6c 24 10 eb 23 66 0f 1f 44 00 00 4c 89 e7 49 c7 
[  229.820052] RIP  [<ffffffff8170ecb0>] __mutex_lock_slowpath+0xb0/0x1f0
[  229.820052]  RSP <ffff88001b5cfdc0>
[  229.820052] CR2: 0000000000000000
[  229.956257] ---[ end trace 582d0bfbf7f0ff79 ]---

Error 2 (normal over shim eth). Start both ipcms, enroll and register an echo-time server in one side. Run 4 parallel echo-time-clients and kill the IPCM where the echo-time-clients are running.

[  980.876114] BUG: unable to handle kernel NULL pointer dereference at           (null)
[  980.880044] IP: [<ffffffff81376a00>] strcmp+0x20/0x40
[  980.880044] PGD 7c1dd067 PUD 29c8c067 PMD 0 
[  980.880044] Oops: 0000 [#1] SMP 
[  980.880044] Modules linked in: normal_ipcp rina_default_plugin shim_eth_vlan rinarp arp826 rina_personality_default 8021q garp stp llc nfsd auth_rpcgss oid_registry nfs_acl nfs lockd fscache sunrpc psmouse pcspkr serio_raw i2c_piix4 hid_generic floppy usbmouse usbhid hid
[  980.880044] CPU: 0 PID: 5199 Comm: kworker/u31:1 Not tainted 3.15.0 #22
[  980.880044] Hardware name: Xen HVM domU, BIOS 4.1.4 08/17/2014
[  980.880044] Workqueue: shim-eth-vlan rwq_worker
[  980.880044] task: ffff88002f1bc8c0 ti: ffff880060604000 task.ti: ffff880060604000
[  980.880044] RIP: 0010:[<ffffffff81376a00>]  [<ffffffff81376a00>] strcmp+0x20/0x40
[  980.880044] RSP: 0018:ffff880060607d68  EFLAGS: 00010202
[  980.880044] RAX: 0000000000000074 RBX: 0000000000000000 RCX: 0000000000000000
[  980.880044] RDX: ffffffffa01c1040 RSI: 0000000000000001 RDI: ffff880030249841
[  980.880044] RBP: ffff880060607d68 R08: ffff880060604000 R09: 0000000000000001
[  980.880044] R10: 0000000000000146 R11: 0000000000000002 R12: ffff88002f05c1a0
[  980.880044] R13: ffff88002f05ca60 R14: ffff88002f05cae0 R15: ffff8800607ad8f0
[  980.880044] FS:  0000000000000000(0000) GS:ffff88007fc00000(0000) knlGS:0000000000000000
[  980.880044] CS:  0010 DS: 0000 ES: 0000 CR0: 000000008005003b
[  980.880044] CR2: 0000000000000000 CR3: 000000003017b000 CR4: 00000000001006f0
[  980.880044] Stack:
[  980.880044]  ffff880060607da8 ffffffffa019b2ea ffff8800607adc00 ffff88006072a780
[  980.880044]  ffff88002f05cae0 ffff880061c19b00 ffff880029fa2c00 00000000000003e0
[  980.880044]  ffff880060607dc8 ffffffffa019d1d2 ffff88002fae90c0 ffff88002c7f0640
[  980.880044] Call Trace:
[  980.880044]  [<ffffffffa019b2ea>] ipcp_imap_find_by_name+0x7a/0xe0 [rina_personality_default]
[  980.880044]  [<ffffffffa019d1d2>] kipcm_find_ipcp_by_name+0x32/0x60 [rina_personality_default]
[  980.880044]  [<ffffffffa01b610d>] eth_vlan_rcv_worker+0x3d/0x2f0 [shim_eth_vlan]
[  980.880044]  [<ffffffff816ebc9b>] rwq_worker+0x1b/0x90
[  980.880044]  [<ffffffff81082052>] process_one_work+0x182/0x4e0
[  980.880044]  [<ffffffff810839ec>] worker_thread+0x13c/0x4c0
[  980.880044]  [<ffffffff810838b0>] ? manage_workers.isra.25+0x360/0x360
[  980.880044]  [<ffffffff8108a1bb>] kthread+0xdb/0x100
[  980.880044]  [<ffffffff8108a0e0>] ? kthread_create_on_node+0x180/0x180
[  980.880044]  [<ffffffff81719dfc>] ret_from_fork+0x7c/0xb0
[  980.880044]  [<ffffffff8108a0e0>] ? kthread_create_on_node+0x180/0x180
[  980.880044] Code: 66 66 2e 0f 1f 84 00 00 00 00 00 55 48 89 e5 eb 0e 66 2e 0f 1f 84 00 00 00 00 00 84 c0 74 1c 48 83 c7 01 0f b6 47 ff 48 83 c6 01 <3a> 46 ff 74 eb 19 c0 83 c8 01 5d c3 0f 1f 40 00 31 c0 5d c3 66 
[  980.880044] RIP  [<ffffffff81376a00>] strcmp+0x20/0x40
[  980.880044]  RSP <ffff880060607d68>
[  980.880044] CR2: 0000000000000000
[  980.880044] ---[ end trace c67c02b1a9796ec0 ]---

sandervrijders commented:
I fixed error 1, that bug is also present in pristine-1.3, but doesn't surface there it seems. Give it another go if possible. I also added some logs for convenience's sake.

sandervrijders commented:
Error 2 is not due to this PR as far as I can tell. Seems related to the name being freed while another thread is in imap_entry_find_by_name (in kipcm-utils.c) which calls strcmp on that name. So it's timing related. Locking in the KIPCM is not the problem however. The problem is that the name of the IPCP is freed by the IPCP itself while it is also passed as a parameter to the KIPCM. Taking a copy of the string may be a solution.

Sidenote: I don't know why hash_for_each_safe is used in that function, since it is only needed if you delete an entry, not when looking it up.

edugrasa commented:
Verified that error 1 is now fixed. Going to test a little bit more!

edugrasa commented:
Got another bang (checking if can also happen in 1.3). Scenario is a DIF with 3 IPCPs over a DIF two shim-eth-vlans. Rina-tgen client sending to a rina-tgen server. I killed the IPCM of the IPCP in the middle while the rina-tgen transfer was going on, and got the following error:

[24883.920121] rina-kipcm(ERR): Couldn't find the ipc process 2
[24883.929379] rina-shim-eth-vlan(ERR): Could not find required ipcps
[24883.929456] rina-rmt(ERR): Could not find the N-1 port
[24883.929457] rina-normal-ipc(ERR): Could not send to RMT (using port_id)
[24883.952926] general protection fault: 0000 [#1] SMP 
[24883.956891] Modules linked in: rina_default_plugin shim_eth_vlan rinarp arp826 nfsd auth_rpcgss oid_registry nfs_acl nfs lockd fscache sunrpc normal_ipcp rina_personality_default 8021q garp stp llc psmouse i2c_piix4 pcspkr serio_raw hid_generic floppy usbmouse usbhid hid
[24883.956891] CPU: 0 PID: 5419 Comm: kworker/u31:1 Not tainted 3.15.0 #8
[24883.956891] Hardware name: Xen HVM domU, BIOS 4.1.4 08/17/2014
[24883.956891] Workqueue: shim-eth-vlan rwq_worker
[24883.956891] task: ffff88002e540000 ti: ffff88003dbc4000 task.ti: ffff88003dbc4000
[24883.956891] RIP: 0010:[<ffffffff811a3ef6>]  [<ffffffff811a3ef6>] kfree+0x56/0x150
[24883.956891] RSP: 0018:ffff88003dbc7d08  EFLAGS: 00010207
[24883.956891] RAX: 037ab5e000004000 RBX: dead000000100100 RCX: 00000000000c0f87
[24883.956891] RDX: 000077ff80000000 RSI: ffff88003fc17640 RDI: dead000000100100
[24883.956891] RBP: ffff88003dbc7d20 R08: 0000000000017640 R09: 037a9fe000004000
[24883.956891] R10: ffffffff816eb52e R11: ffff8800198a5000 R12: ffff88003e3a21e0
[24883.956891] R13: ffffffff816eb52e R14: ffffffff816f39e0 R15: 00000000000003e0
[24883.956891] FS:  0000000000000000(0000) GS:ffff88003fc00000(0000) knlGS:0000000000000000
[24883.956891] CS:  0010 DS: 0000 ES: 0000 CR0: 000000008005003b
[24883.956891] CR2: 0000000008f019c8 CR3: 000000002a153000 CR4: 00000000001006f0
[24883.956891] Stack:
[24883.956891]  ffff88002a13d7c0 ffff88003e3a21e0 ffff88003e3a2c20 ffff88003dbc7d30
[24883.956891]  ffffffff816eb52e ffff88003dbc7d40 ffffffff816f3403 ffff88003dbc7d58
[24883.956891]  ffffffff816f39ff ffff88003e3a2ea0 ffff88003dbc7d88 ffffffff816ec291
[24883.956891] Call Trace:
[24883.956891]  [<ffffffff816eb52e>] rkfree+0xe/0x10
[24883.956891]  [<ffffffff816f3403>] pci_destroy+0x13/0x30
[24883.956891]  [<ffffffff816f39ff>] pdu_destroy+0x1f/0x40
[24883.956891]  [<ffffffff816ec291>] rqueue_destroy+0x61/0xb0
[24883.956891]  [<ffffffff816ec491>] rfifo_destroy+0x21/0x70
[24883.956891]  [<ffffffffa01d1662>] flow_destroy+0x82/0xa0 [shim_eth_vlan]
[24883.956891]  [<ffffffffa01d224b>] eth_vlan_rcv_worker+0x17b/0x2f0 [shim_eth_vlan]
[24883.956891]  [<ffffffff816ebc9b>] rwq_worker+0x1b/0x90
[24883.956891]  [<ffffffff81082052>] process_one_work+0x182/0x4e0
[24883.956891]  [<ffffffff810839ec>] worker_thread+0x13c/0x4c0
[24883.956891]  [<ffffffff810838b0>] ? manage_workers.isra.25+0x360/0x360
[24883.956891]  [<ffffffff8108a1bb>] kthread+0xdb/0x100
[24883.956891]  [<ffffffff8108a0e0>] ? kthread_create_on_node+0x180/0x180
[24883.956891]  [<ffffffff81719dfc>] ret_from_fork+0x7c/0xb0
[24883.956891]  [<ffffffff8108a0e0>] ? kthread_create_on_node+0x180/0x180
[24883.956891] Code: 00 00 00 80 ff 77 00 00 49 b9 00 00 00 00 00 ea ff ff 48 01 d8 48 0f 42 15 28 41 a7 00 48 01 d0 48 c1 e8 0c 48 c1 e0 06 49 01 c1 <49> 8b 01 f6 c4 80 0f 85 b6 00 00 00 49 8b 01 a8 80 0f 84 83 00 
[24883.956891] RIP  [<ffffffff811a3ef6>] kfree+0x56/0x150
[24883.956891]  RSP <ffff88003dbc7d08>
[24884.272084] ---[ end trace ecb34956180077c7 ]---
[24884.277668] BUG: unable to handle kernel paging request at ffffffffffffffd8
[24884.281632] IP: [<ffffffff8108a970>] kthread_data+0x10/0x20
[24884.281632] PGD 1c14067 PUD 1c16067 PMD 0 
[24884.281632] Oops: 0000 [#2] SMP 
[24884.281632] Modules linked in: rina_default_plugin shim_eth_vlan rinarp arp826 nfsd auth_rpcgss oid_registry nfs_acl nfs lockd fscache sunrpc normal_ipcp rina_personality_default 8021q garp stp llc psmouse i2c_piix4 pcspkr serio_raw hid_generic floppy usbmouse usbhid hid
[24884.281632] CPU: 0 PID: 5419 Comm: kworker/u31:1 Tainted: G      D       3.15.0 #8
[24884.281632] Hardware name: Xen HVM domU, BIOS 4.1.4 08/17/2014
[24884.281632] task: ffff88002e540000 ti: ffff88003dbc4000 task.ti: ffff88003dbc4000
[24884.281632] RIP: 0010:[<ffffffff8108a970>]  [<ffffffff8108a970>] kthread_data+0x10/0x20
[24884.281632] RSP: 0018:ffff88003dbc7ac8  EFLAGS: 00010002
[24884.281632] RAX: 0000000000000000 RBX: 0000000000000000 RCX: 0000000000000000
[24884.281632] RDX: 000000000000000f RSI: 0000000000000000 RDI: ffff88002e540000
[24884.281632] RBP: ffff88003dbc7ac8 R08: 0000000000000000 R09: 0000000000000001
[24884.281632] R10: 0000000000001efe R11: 0000000000000000 R12: ffff88003fc14840
[24884.281632] R13: ffff88003fc14840 R14: ffff88002e540000 R15: 0000000000000000
[24884.281632] FS:  0000000000000000(0000) GS:ffff88003fc00000(0000) knlGS:0000000000000000
[24884.281632] CS:  0010 DS: 0000 ES: 0000 CR0: 000000008005003b
[24884.281632] CR2: 0000000000000028 CR3: 000000002a153000 CR4: 00000000001006f0
[24884.281632] Stack:
[24884.281632]  ffff88003dbc7ae0 ffffffff81083df1 0000000000000000 ffff88003dbc7b40
[24884.281632]  ffffffff8170c0aa ffff88002e540000 0000000000014840 ffff88003dbc7fd8
[24884.281632]  0000000000014840 ffff88002e540000 ffff88002e540748 ffff88002e53fff0
[24884.281632] Call Trace:
[24884.281632]  [<ffffffff81083df1>] wq_worker_sleeping+0x11/0x90
[24884.281632]  [<ffffffff8170c0aa>] __schedule+0x6ba/0x850
[24884.281632]  [<ffffffff8170c269>] schedule+0x29/0x70
[24884.281632]  [<ffffffff810678d8>] do_exit+0x818/0xac0
[24884.281632]  [<ffffffff81711f3d>] oops_end+0xad/0x150
[24884.281632]  [<ffffffff81016f5b>] die+0x4b/0x70
[24884.281632]  [<ffffffff816f39e0>] ? pdu_buffer_set+0x40/0x40
[24884.281632]  [<ffffffff81711982>] do_general_protection+0xd2/0x160
[24884.281632]  [<ffffffff816eb52e>] ? rkfree+0xe/0x10
[24884.281632]  [<ffffffff817112e8>] general_protection+0x28/0x30
[24884.281632]  [<ffffffff816f39e0>] ? pdu_buffer_set+0x40/0x40
[24884.281632]  [<ffffffff816eb52e>] ? rkfree+0xe/0x10
[24884.281632]  [<ffffffff816eb52e>] ? rkfree+0xe/0x10
[24884.281632]  [<ffffffff811a3ef6>] ? kfree+0x56/0x150
[24884.281632]  [<ffffffff816eb52e>] rkfree+0xe/0x10
[24884.281632]  [<ffffffff816f3403>] pci_destroy+0x13/0x30
[24884.527393] rina-rnl(WARN): IPC Manager process has been destroyed
[24884.281632]  [<ffffffff816f39ff>] pdu_destroy+0x1f/0x40
[24884.281632]  [<ffffffff816ec291>] rqueue_destroy+0x61/0xb0
[24884.281632]  [<ffffffff816ec491>] rfifo_destroy+0x21/0x70
[24884.281632]  [<ffffffffa01d1662>] flow_destroy+0x82/0xa0 [shim_eth_vlan]
[24884.281632]  [<ffffffffa01d224b>] eth_vlan_rcv_worker+0x17b/0x2f0 [shim_eth_vlan]
[24884.281632]  [<ffffffff816ebc9b>] rwq_worker+0x1b/0x90
[24884.281632]  [<ffffffff81082052>] process_one_work+0x182/0x4e0
[24884.281632]  [<ffffffff810839ec>] worker_thread+0x13c/0x4c0
[24884.281632]  [<ffffffff810838b0>] ? manage_workers.isra.25+0x360/0x360
[24884.281632]  [<ffffffff8108a1bb>] kthread+0xdb/0x100
[24884.281632]  [<ffffffff8108a0e0>] ? kthread_create_on_node+0x180/0x180
[24884.281632]  [<ffffffff81719dfc>] ret_from_fork+0x7c/0xb0
[24884.281632]  [<ffffffff8108a0e0>] ? kthread_create_on_node+0x180/0x180
[24884.281632] Code: 00 48 89 e5 5d 48 8b 40 c8 48 c1 e8 02 83 e0 01 c3 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 44 00 00 48 8b 87 d0 04 00 00 55 48 89 e5 <48> 8b 40 d8 5d c3 66 2e 0f 1f 84 00 00 00 00 00 0f 1f 44 00 00 
[24884.281632] RIP  [<ffffffff8108a970>] kthread_data+0x10/0x20
[24884.281632]  RSP <ffff88003dbc7ac8>
[24884.281632] CR2: ffffffffffffffd8
[24884.281632] ---[ end trace ecb34956180077c8 ]---
[24884.281632] Fixing recursive fault but reboot is needed!

sandervrijders commented:
I fixed your post to use ` ` ` around code quotation. That error seems related to the previous one also, IPCPs are being destroyed while they are still doing stuff.

edugrasa commented:
@sandervrijders tried with pristine-1.3, error 3 also happens (so it's not the fault of this PR). I'll open a new ticket so that we don't forget. Tomorrow morning I'm going to try normal over normal and see if it also works.

edugrasa commented:
ACK, normal over normal works as well as pristine-1.3, I didn't found extra bugs in addition of the two that are not RMT-related and have already been reported.

lbergesio commented:
I remove my NACK at the beggining of the discussion since that was fixed, so ACK too
